src/gradient.toit

```diff
@@ -0,0 +1,303 @@
+      texture_length_ = texture_length
+
+      red_pixels_ = ByteArray texture_length * texture_length_repeats_
+      green_pixels_ = ByteArray texture_length * texture_length_repeats_
+      blue_pixels_ = ByteArray texture_length * texture_length_repeats_
+      ranges/List := extract_ranges_ specifiers
+      ranges.do: | range |
+        get_colors range texture_length: | index red green blue |
+          red_pixels_[index] = red
+          green_pixels_[index] = green
+          blue_pixels_[index] = blue
+        (texture_length_repeats_ - 1).repeat:
+          red_pixels_.replace ((it + 1) * texture_length) red_pixels_[0..texture_length]
+          green_pixels_.replace ((it + 1) * texture_length) green_pixels_[0..texture_length]
+          blue_pixels_.replace ((it + 1) * texture_length) blue_pixels_[0..texture_length]
+
+  draw canvas/Canvas x/int y/int --autocropped/bool -> none:
+    if not canvas.supports_8_bit: throw "UNSUPPORTED"
+    angle := angle_
+    w := w_
+    h := h_
+    analysis := canvas.bounds_analysis x y w h
+    if analysis == Canvas.DISJOINT: return
+    // Determine whether the draw operations will be automatically cropped for
+    // us, or whether we need to do it ourselves by using slices for drawing
+    // operations.  We could also check whether we are inside a window that will
+    // use compositing to crop everything.
+    if analysis == Canvas.CANVAS_IN_AREA or analysis == Canvas.COINCIDENT: autocropped = true
+
+    // CSS gradient angles are:
+    //    0 bottom to top.
+    //   90 left to right
+    //  180 top to bottom
+    //  270 right to left
+
+    repeats := texture_length_repeats_
+    offset := 0
+    source_width := 0
+    r := null
+    g := null
+    b := null
+
+    draw_block := : | y2 hw lines orientation |
+      o := offset >> 16
+      // Repeats != 1 implies the gradient is vertical or horizontal, which
+      // means we don't need clipping.
+      if autocropped or repeats != 1:
+        source_width = texture_length_
+        buf_size := texture_length_ * lines
+        r = red_pixels_[..buf_size]
+        g = green_pixels_[..buf_size]
+        b = blue_pixels_[..buf_size]
+        if orientation == ORIENTATION_90:
+          y2 + o  // Return value.
+        else:
+          y2 - o  // Return value.
+      else:
+        r = red_pixels_[o .. o + hw]
+        g = green_pixels_[o .. o + hw]
+        b = blue_pixels_[o .. o + hw]
+        source_width = hw
+        y2  // Return value.
+
+    if draw_vertical_:
+      // The gradient goes broadly vertically, and we draw in vertical strips.
+      orientation/int := ORIENTATION_90
+      x2/int := x
+      y2/int := y + h
+      if angle >= 90:  // Top to bottom.
+        orientation = ORIENTATION_270
+        x2++
+        y2 = y
+      step := ((texture_length_ - h) << 16) / w  // n.16 fixed point.
+      for i := 0; i < w; i += repeats:
+        lines := min repeats (w - i)
+        y3 := draw_block.call y2 h lines orientation
```

Comment:
too much. we now have `y`, `y2` and `y3`.
Maybe it's ok if each version has a comment what it is.

Reply:
I added some comments and renamed some variables.